third_party/blink/renderer/modules/cobalt/h5vcc_experiments/h_5_vcc_experiments.cc

```diff
@@ -42,58 +43,18 @@ ScriptPromise H5vccExperiments::setExperimentState(
 
   EnsureReceiverIsBound();
 
-  if (!experiment_configuration->hasFeatures() ||
-      !experiment_configuration->hasFeatureParams() ||
-      !experiment_configuration->hasExperimentIds()) {
-    resolver->RejectWithDOMException(DOMExceptionCode::kInvalidStateError,
-                                     "Incomplete experiment configuration.");
-    return promise;
-  }
+  std::optional<base::Value::Dict> experiment_config_dict =
+      ToBaseValueDict(experiment_configuration);
 
-  base::Value::Dict experiment_config_dict;
-
-  base::Value::Dict features;
-  for (auto& feature_name_and_value : experiment_configuration->features()) {
-    features.Set(feature_name_and_value.first.Utf8(),
-                 feature_name_and_value.second);
-  }
-  experiment_config_dict.Set(cobalt::kExperimentConfigFeatures,
-                             std::move(features));
-
-  // All FieldTrialParams are stored as strings, including booleans.
-  base::Value::Dict feature_params;
-  std::string param_value;
-  for (auto& param_name_and_value : experiment_configuration->featureParams()) {
-    if (param_name_and_value.second->IsString()) {
-      param_value = param_name_and_value.second->GetAsString().Utf8();
-    } else if (param_name_and_value.second->IsLong()) {
-      param_value = std::to_string(param_name_and_value.second->GetAsLong());
-    } else if (param_name_and_value.second->GetAsBoolean()) {
-      param_value = "true";
-    } else if (!param_name_and_value.second->GetAsBoolean()) {
-      param_value = "false";
-    } else {
-      // TODO (b/416323107) - In thoery this never gets executed because
-      // wrong types are implicitly converted. We need to enforce the typing
-      // from the callsite.
-      resolver->RejectWithDOMException(DOMExceptionCode::kInvalidStateError,
-                                       "Invalid feature param type.");
-      return promise;
-    }
-    feature_params.Set(param_name_and_value.first.Utf8(), param_value);
-  }
-  experiment_config_dict.Set(cobalt::kExperimentConfigFeatureParams,
-                             std::move(feature_params));
-
-  base::Value::List experiment_ids;
-  for (int exp_id : experiment_configuration->experimentIds()) {
-    experiment_ids.Append(exp_id);
+  if (!experiment_config_dict.has_value()) {
+    resolver->RejectWithDOMException(
+        DOMExceptionCode::kInvalidStateError,
+        "Unable to parse experiment configuration.");
```

Comment:
I'm wondering about this rejection... it basically means any caller of `setExperimentState` _must_ call it with a complete experiment_configuration object otherwise the promise fails... But what if, for example, there really are no feature params or experiment IDs? It feels odd the call would reject in this case.

Moreover, your IDL definition doesn't mark the ExperimentConfiguration properties as `required`. So you effectively treat them as required, but the IDL doesn't indicate that. 

I'm thinking we need to consider the properties optional here... and just assume it's an empty set if they're missing.